File: 0050-powx-n/0050-powx-n.go
func myPow(x float64, n int) float64 {
    if n == 0 {
        return 1
    }
    
    negExp := false
    if n < 0 {
        n = -n
        negExp = true
    }
    
    result := float64(1)
    for n > 0 {
        scale := 1
        multiplier := x
        for scale * 2 < n {
            scale *= 2
            multiplier *= multiplier
        }
        
        result *= multiplier
        n -= scale
    }
    
    if negExp {
        return 1 / result
    }
    return result
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0: 
            return 0
        
        neg_exp = False
        if n < 0:
            n = -n
            neg_exp = True
        
        res = 1
        while n > 0:
            scale = 1
            new_x = x
            while scale * 2 < n:
                new_x *= new_x
                scale *= 2

            n -= scale
            res *= new_x
                
        if neg_exp:
            return 1 / res
        
        return res
        